### ISSUE TITLE
x0vncserver does not support clipboard

### ISSUE THREAD

**Issue by astrand** [TigerVNC/tigervnc#529]
Currently, x0vncserver has no support for sending clipboard changes to the client, or accept clipboard updates from the client.

**Comment by Racansky** [comment]
Hi,
for x0vncserver I used to run the parameters `-SendCutText -AcceptCutText` - then my clipboard works in both directions

**Comment by CendioOssman** [comment]
I'm afraid that can't be our `x0vncserver` in that case. This is the complete current code for clipboard handling right now:

```C++
void XDesktop::clientCutText(const char* str, int len) {
}
```

:)

**Comment by TheQuantumPhysicist** [comment]
I thought I found the perfect VNC server... but now this... I seriously can't catch a break with VNC!

**Comment by Mannshoch** [comment]
Is that problem still available?

**Comment by CendioOssman** [comment]
I'm afraid no one has had a look at this yet.

**Comment by ivorob** [comment]
Hi. Is the bounty active for that issue? I want to try resolve it to start working with project.

**Comment by CendioOssman** [comment]
Yes, the bounty is still active so feel free to start hacking. :)

**Comment by ivorob** [comment]
Guys, I am close to provide code to code review. But I have a several questions. Do you have some code style standard? Where can I find it. Sorry for offtopic.

**Comment by CendioOssman** [comment]
Sorry, no. You'll have to look at existing files and try to follow the same style.

**Comment by mathewng** [comment]
After establishing a connection to `X0tigervnc`, I launched `x11vnc -nofb -clip 4x4+0+0` to synchronise the clipboard.
This works for me.

**Comment by markd89** [comment]
@mathewng can you post more details of your workaround. When I try that x11vnc listens on 5901 and I don't get any clipboard transfer in tigervncviewer. I'm thinking that maybe I need to run a second viewer to get the clipboard transfer?

@ivorob how are you doing with your code? You may be able to find something in the x11vnc source code that you can modify to add to x0vncserver. I would be willing to add to the bountysource if you could get this done soon.

**Comment by mathewng** [comment]
> @mathewng can you post more details of your workaround. When I try that x11vnc listens on 5901 and I don't get any clipboard transfer in tigervncviewer. I'm thinking that maybe I need to run a second viewer to get the clipboard transfer?
> 
> @ivorob how are you doing with your code? You may be able to find something in the x11vnc source code that you can modify to add to x0vncserver. I would be willing to add to the bountysource if you could get this done soon.

Yes, you are right. 

My script to launch vnc server goes roughly like this
It works though there is a noticeable delay in the clipboard syncing when copying and pasting. not copy & paste but copy, wait and paste

```
SSH_CLIENT=...

# For display
x0tigervncserver \
-display :0 \
-localhost yes \
-useold

echo starting clipboard sync in 20 seconds
sleep 20

# For clipboard sync
sudo /usr/bin/x11vnc \
-display :0 \
-auth "${DISPLAYMANAGER_AUTH}" \
-noipv6 \
-connect_or_exit "${SSH_CLIENT%% *}:5500" \
-users meofcourse \
-nopw \
-noremote \
-nocmds \
-xrandr resize \
-xinerama \
-nomodtweak \
-clear_mods \
-skip_lockkeys \
-once \
-noxfixes \
-cursor_drag \
-repeat \
-allinput \
-bg \
-nofb \
-clip 4x4+0+0
```

I connect to `x0tigervncserver` for display
and have turbovncviewer in listen mode ready for `x11vnc` to connect to it

**Comment by wanderling** [comment]
Putting my money where my mouth is. Added $20 to this bounty. [Bounty now $50](https://app.bountysource.com/issues/50701623-x0vncserver-does-not-support-clipboard?utm_campaign=plugin&utm_content=tracker%2F3557444&utm_medium=issues&utm_source=github)

**Comment by ValkaTR** [comment]
> Putting my money where my mouth is. Added $20 to this bounty. [Bounty now $50](https://app.bountysource.com/issues/50701623-x0vncserver-does-not-support-clipboard?utm_campaign=plugin&utm_content=tracker%2F3557444&utm_medium=issues&utm_source=github)

+ $25

**Comment by mvalley21** [comment]
I also threw $25 into the bounty pot. This would be a useful feature for many.

**Comment by mvalley21** [comment]
BTW, my workaround for now is using this program on the client and server: https://github.com/megahertz/clipboard-share

**Comment by bitcoinmeetups** [comment]
I second this. I can't believe it's been 7 years since this post and x0vncserver still can't support copy and paste from connecting unit. Come on.

**Comment by bitcoinmeetups** [comment]
How can we even use a vnc program if clipbaoard is not supported? What is the purpose of the program then? To watch the Terminal from a distance without being able to interact with it normally?

**Comment by bitcoinmeetups** [comment]
I also thought I had found the final solution for all the enormous vnc hassle. And then this.

**Comment by Skgland** [comment]
Note that to my knowledge this only applies for the x0vncserver not the x11 plugin (libvnc.so) [see "3 Expose the local display directly"](https://wiki.archlinux.org/title/TigerVNC#Expose_the_local_display_directly) or the standalone Xvnc server.

I am currently using the libvnc.so with working copy-paste at work.

**Comment by TTimo** [comment]
Remmina has 'Send clipboard content as keystrokes' which I use as a workaround. For larger pastes I've used cryptpad or a text file on dropbox between systems.

**Comment by CendioOssman** [comment]
This has now been implemented via #1831. Many thanks to @gujjwal00!

**Comment by markd89** [comment]
@francoisp This was long a longstanding issue. However, it was supposed to be already implemented. See @CendioOssman comment of September 27, 2024. 

I haven't personally implemented the x0vncserver from what I am using, but it's on my list to do so. 

Still nice to have your contribution as it's another way to do it and more code for learning too.
